{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOwxOuhESW+tPebPucGDEEw"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["Prepare some files so we can simulate error"],"metadata":{"id":"jn1kl96UQGmH"}},{"cell_type":"code","source":["%%writefile test_file.txt\n","Test File Contents!"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"nxPcxuQeQWVH","executionInfo":{"status":"ok","timestamp":1750272183677,"user_tz":240,"elapsed":15,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"650af9da-b2d2-4e5f-a877-8cb0592aaeaf"},"execution_count":54,"outputs":[{"output_type":"stream","name":"stdout","text":["Overwriting test_file.txt\n"]}]},{"cell_type":"markdown","source":["If we mess up with file name when opening it\n","\n","    f = open(\"testfile.txt\")\n","\n","we get an error\n","\n","    ---------------------------------------------------------------------------\n","    FileNotFoundError                         Traceback (most recent call last)\n","    /tmp/ipython-input-31-1085390797.py in <cell line: 0>()\n","    ----> 1 f = open(\"testfile.txt\")\n","\n","    FileNotFoundError: [Errno 2] No such file or directory: 'testfile.txt'"],"metadata":{"id":"ria_o5A-QlY8"}},{"cell_type":"markdown","source":["#### Try/except"],"metadata":{"id":"ZHoI7jIwS9Wl"}},{"cell_type":"markdown","source":["So, if we can anticipate such error might happen and we want to handle it more gracefully, try/except block can help"],"metadata":{"id":"XfewU9bJQ1G4"}},{"cell_type":"code","source":["try:\n","  f = open(\"testfile.txt\")\n","  var = bad_var\n","except FileNotFoundError as e:\n","  print(e) # [Errno 2] No such file or directory: 'testfile.py'\n","except NameError as e:\n","  print(e) # name 'bad_var' is not defined\n","except Exception as e:\n","  print(e) # name 'bad_var' is not defined"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"DQDaATo3Q7gF","executionInfo":{"status":"ok","timestamp":1750272183682,"user_tz":240,"elapsed":4,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"e32dbf68-343a-4252-ca87-f246466b0219"},"execution_count":55,"outputs":[{"output_type":"stream","name":"stdout","text":["[Errno 2] No such file or directory: 'testfile.txt'\n"]}]},{"cell_type":"markdown","source":["We see, we only show custom error message rather than standard error trace"],"metadata":{"id":"QnmzZQ5mRFfi"}},{"cell_type":"markdown","source":["#### Else clause"],"metadata":{"id":"63bnnTSnS_FE"}},{"cell_type":"markdown","source":["This is triggerred if `try` does not throw any exception"],"metadata":{"id":"mkXbRj6MTAeK"}},{"cell_type":"code","source":["try:\n","  f = open(\"testfile.txt\")\n","except FileNotFoundError as e:\n","  print(e) # [Errno 2] No such file or directory: 'testfile.py'\n","else:\n","  print(\"No exception thrown, else clause runs\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"X1-_XAtOREmT","executionInfo":{"status":"ok","timestamp":1750272183687,"user_tz":240,"elapsed":5,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"9fc927fe-6a16-46d2-8420-21dd9b4456bc"},"execution_count":56,"outputs":[{"output_type":"stream","name":"stdout","text":["[Errno 2] No such file or directory: 'testfile.txt'\n"]}]},{"cell_type":"code","source":["try:\n","  f = open(\"test_file.txt\")\n","except FileNotFoundError as e:\n","  print(e) # [Errno 2] No such file or directory: 'testfile.py'\n","else:\n","  print(\"No exception thrown, else clause runs\")\n","  print(f.read())\n","  f.close()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"TE6x8W4nTLGu","executionInfo":{"status":"ok","timestamp":1750272183695,"user_tz":240,"elapsed":8,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"9e20145c-3281-486a-c21f-7f25d888368a"},"execution_count":57,"outputs":[{"output_type":"stream","name":"stdout","text":["No exception thrown, else clause runs\n","Test File Contents!\n","\n"]}]},{"cell_type":"markdown","source":["#### Finally clause"],"metadata":{"id":"-xi2ik4WUMqu"}},{"cell_type":"markdown","source":["It is triggered no matter whether `try` throws exception or not\n","\n","Can be useful when we want to release certain resources whether try is successful or not"],"metadata":{"id":"CmTsbvDeUOlO"}},{"cell_type":"code","source":["try:\n","  f = open(\"testfile.txt\")\n","except FileNotFoundError as e:\n","  print(e) # [Errno 2] No such file or directory: 'testfile.py'\n","else:\n","  print(\"No exception thrown, else clause runs\")\n","finally:\n","  print(\"Finally clause always runs\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"oM8hdDkBTOKa","executionInfo":{"status":"ok","timestamp":1750272183711,"user_tz":240,"elapsed":15,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"32ae7e70-1a12-406e-c179-a3c319f97833"},"execution_count":58,"outputs":[{"output_type":"stream","name":"stdout","text":["[Errno 2] No such file or directory: 'testfile.txt'\n","Finally clause always runs\n"]}]},{"cell_type":"code","source":["try:\n","  f = open(\"test_file.txt\")\n","except FileNotFoundError as e:\n","  print(e) # [Errno 2] No such file or directory: 'testfile.py'\n","else:\n","  print(\"No exception thrown, else clause runs\")\n","finally:\n","  print(\"Finally clause always runs\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"JcFe29NnUql_","executionInfo":{"status":"ok","timestamp":1750272183714,"user_tz":240,"elapsed":2,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"ab36092f-96b7-4267-eed9-a89c0281d0c1"},"execution_count":59,"outputs":[{"output_type":"stream","name":"stdout","text":["No exception thrown, else clause runs\n","Finally clause always runs\n"]}]},{"cell_type":"markdown","source":["#### Manual exception"],"metadata":{"id":"8gkJRTTZVcE-"}},{"cell_type":"markdown","source":["We can also manually define exception condition"],"metadata":{"id":"dFZ_0t8fVe--"}},{"cell_type":"code","source":["try:\n","  f = open(\"test_file.txt\")\n","  if f.name == \"testfile.txt\":\n","    raise Exception\n","except FileNotFoundError as e:\n","  print(e) # [Errno 2] No such file or directory: 'testfile.py'\n","except Exception as e:\n","  print('Customized exception!')\n","else:\n","  print(\"No exception thrown, else clause runs\")\n","finally:\n","  print(\"Finally clause always runs\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"VfHa9G0aUrNF","executionInfo":{"status":"ok","timestamp":1750272183730,"user_tz":240,"elapsed":15,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"fef246f6-30c7-4590-fcd0-e6c1aac1c26d"},"execution_count":60,"outputs":[{"output_type":"stream","name":"stdout","text":["No exception thrown, else clause runs\n","Finally clause always runs\n"]}]},{"cell_type":"code","source":["try:\n","  f = open(\"test_file.txt\")\n","  if f.name == \"test_file.txt\":\n","    raise Exception\n","except FileNotFoundError as e:\n","  print(e) # [Errno 2] No such file or directory: 'testfile.py'\n","except Exception as e:\n","  print('Customized exception!')\n","else:\n","  print(\"No exception thrown, else clause runs\")\n","finally:\n","  print(\"Finally clause always runs\")"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"7_bMEbQSVsCu","executionInfo":{"status":"ok","timestamp":1750272183732,"user_tz":240,"elapsed":3,"user":{"displayName":"Jianjing Zhang","userId":"07583447684872889447"}},"outputId":"129ef178-3457-447d-a6a5-3f1ae37c337f"},"execution_count":61,"outputs":[{"output_type":"stream","name":"stdout","text":["Customized exception!\n","Finally clause always runs\n"]}]}]}